src/detectors/BTOF/BarrelTOFNeighborFinder.cc

```diff
@@ -0,0 +1,201 @@
+  if (!_converter)
+    throw std::runtime_error(
+        "CellIDPositionConverter not initialized in BarrelTOFNeighborFinder. Have you called "
+        "BarrelTOFNeighborFinder::init(/* detector */) first?");
+  return _converter->cellID(pos);
+}
+
+dd4hep::Position
+BarrelTOFNeighborFinder::local2GlobalInStaveFromCell(const dd4hep::rec::CellID& cell,
+                                                     const dd4hep::Position& pos) {
+  // convert local position to global position
+  // assuming the local position is located at the same volume as cell
+  if (!_detector)
+    throw std::runtime_error("Detector ptr not initialized in BarrelTOFNeighborFinder. Have you "
+                             "called BarrelTOFNeighborFinder::init(/* detector */) first?");
+
+  auto geoManager = _detector->world().volume()->GetGeoManager();
+  auto position   = this->cell2GlobalPosition(cell);
+  auto node       = geoManager->FindNode(position.x(), position.y(), position.z());
+  auto currMatrix = geoManager->GetCurrentMatrix();
+
+  double g[3], l[3];
+  pos.GetCoordinates(l);
+  currMatrix->LocalToMaster(l, g); // <- only different from cell2LocalPosition
+  position.SetCoordinates(g);      // <- only different from cell2LocalPosition
+  return position;
+}
+
+dd4hep::Position BarrelTOFNeighborFinder::global2Local(const dd4hep::Position& pos) {
+  // convert local position to global position
+  // assuming the local position is located at the same volume as cell
+  if (!_detector)
+    throw std::runtime_error("Detector ptr not initialized in BarrelTOFNeighborFinder. Have you "
+                             "called BarrelTOFNeighborFinder::init(/* detector */) first?");
+  auto geoManager = _detector->world().volume()->GetGeoManager();
+  auto node       = geoManager->FindNode(pos.x(), pos.y(), pos.z());
+  auto currMatrix = geoManager->GetCurrentMatrix();
+
+  double g[3], l[3];
+  pos.GetCoordinates(g);
+  currMatrix->MasterToLocal(g, l);
+  dd4hep::Position position;
+  position.SetCoordinates(l);
+  return position;
+}
+
+dd4hep::Position BarrelTOFNeighborFinder::cell2LocalPosition(const dd4hep::rec::CellID& cell) {
+  if (!_detector)
+    throw std::runtime_error("Detector ptr not initialized in BarrelTOFNeighborFinder. Have you "
+                             "called BarrelTOFNeighborFinder::init(/* detector */) first?");
+
+  auto geoManager = _detector->world().volume()->GetGeoManager();
+  auto position   = this->cell2GlobalPosition(cell);
+  auto node       = geoManager->FindNode(position.x(), position.y(), position.z());
+  auto currMatrix = geoManager->GetCurrentMatrix();
+
+  double g[3], l[3];
+  position.GetCoordinates(g);
+  currMatrix->MasterToLocal(g, l);
+  position.SetCoordinates(l);
+  return position;
+}
+
+std::vector<double> BarrelTOFNeighborFinder::cellDimension(const dd4hep::rec::CellID& hitCell) {
+  if (!_converter)
+    _initWithCell(hitCell);
+  return _converter->cellDimensions(hitCell);
+}
```

Comment:
The cell dimensions should not change between cells in the same readout, could be defined based on the readout in init. (definitely at least for CartesianGridXY)

Reply:
Ok, looks like something that ideally would be possible in dd4hep but isn't currently and also isn't too important.